### ISSUE TITLE
request object in beforeRequest hook doesnt contain searchParams

### ISSUE THREAD

**Issue by dontwork** [sindresorhus/ky#100]
We are using ky.extend ad have some hooks set up in our instance of ky.

Reading the docs for beforeRequest and it says `Ky will make no further changes to the request after this. ` based on that i expected to be able to access the searchParams that are present for that request but the object looks like this:

```
credentials: "same-origin"
headers: Headers {}
method: "GET"
prefixUrl: "URL"
retry: 2
```

is it possible to access searchParams in this hook?

**Comment by sholladay** [comment]
Hi @dontwork. I looked into this and the problem is essentially that the `options` that Ky passes to your `beforeRequest` hook are _not_ all of the options that the end user passed to Ky, but rather all of the options that Ky will pass to [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch). You seem to expect the former and I can see how this is confusing because `prefixUrl` and `retry` are in there, but that seems like a mistake on our part, since `fetch()` won't know what to do with them and it could lead to problems in the future if new options are added to `fetch()`.

In a similar vein, `fetch()` won't understand `searchParams` and we are not currently passing it, so you don't see it in your `beforeRequest` hook.

We should make this clearer by being more deliberate about it. We want to pass to `fetch()` any options that Ky has normalized plus all unknown options, but probably not any options that are only for Ky itself. That's the set of options that `beforeRequest` should receive before we pass it to `fetch()`.

Once that is done, there's still the issue of accessing the `searchParams` somehow. One idea would be to add a new argument to `beforeRequest` hooks for Ky-specific options. Meh. Or maybe we should pass everything to the hook in one object and then use the hook's return value as the fetch options. There are some different approaches to consider...

Another thing to note is that by the time your hook runs, we've already processed the `searchParams` and modified the request URL accordingly. Your hook runs too late to modify it in any meaningful way even if we passed it to you.

Our documentation could be clearer on all of this, too. It mentions "normalized options" without defining which options and whether they can be modified in the hook. Some can, such as `headers`, while others wouldn't be even if we passed them, such as `searchParams`.

**Comment by dontwork** [comment]
So, what you are saying is that if i wanted to perform some action that added or removed searchParams at a global level this cannot and will not be possible within this hook. The solution would likely be to have a wrapping function around ky have it in there?

**Comment by sholladay** [comment]
> So, what you are saying is that if i wanted to perform some action that added or removed searchParams at a global level this cannot and will not be possible within this hook.

Correct. Reading `searchParams` and taking action based on its value is a possibility we could support in the future if there's a need for it, but this hook will probably never be able to write to it.

One solution for your case might be to add a new type of hook specifically for processing options before Ky does. I'm not sure if this is something we want to support, but I can see it being useful for custom instances. I would like to hear other ideas and opinions.

**Comment by szmarczak** [comment]
Closing in favor of #116
